### PR TITLE
ci: release process for the opbeans-go

### PR DESCRIPTION
**Pull request by v1v**
## Highlights
- Enable ITs: async when running as a PR or sync when running in a branch/tag
- Continuous delivery when running the pipeline either in a branch or tag. master branch refers to `latest` while others will name the image with their names.
- Script to bump versions for the go agent.

## Tasks
- [x] Versioning
- [x] Release automation
- [x] ITs validation should be synchronous when running a release.

## E2E

image

## Issues
- Depends on:
  - https://github.com/elastic/apm-integration-testing/pull/556
  - https://github.com/elastic/apm-pipeline-library/pull/271
  - https://github.com/elastic/apm-agent-go/pull/678

### CODE REVIEW

File: .ci/Jenkinsfile [hunk 1]
```diff
@@ -81,6 +84,19 @@ pipeline {
             }
           }
         }
+        stage('Integration Tests') {
+          agent none
+          steps {
+            log(level: 'INFO', text: 'Launching Async ITs')
+            build(job: env.ITS_PIPELINE, propagate: false, wait: false,
+                  parameters: [string(name: 'AGENT_INTEGRATION_TEST', value: 'Go'),
```
Comment (v1v) [inline .ci/Jenkinsfile]:
```suggestion
                  parameters: [string(name: 'AGENT_INTEGRATION_TEST', value: 'Opbeans'),
```

File: Makefile [hunk 2]
```diff
@@ -28,7 +28,7 @@ test: prepare-test ## Run the tests
 					sh -c "npm install tap-xunit -g && cat target/results.tap | tap-xunit --package='co.elastic.opbeans' > target/junit-results.xml"
 
 publish: build ## Publish docker image
-	@docker push opbeans/opbeans-go:${VERSION}
+	@echo docker push ${IMAGE}:${VERSION}
```
Comment (v1v) [inline Makefile]: For testing purposes only
```suggestion
	@docker push ${IMAGE}:${VERSION}
```

File: .ci/Jenkinsfile [hunk 3]
```diff
@@ -81,24 +81,29 @@ pipeline {
...
-            message 'Should we release a new version?'
-            ok 'Yes, we should.'
-            parameters {
-              string(name: 'VERSION', defaultValue: 'latest', description: 'What tag?')
-            }
+        stage('Integration Tests') {
+          agent none
+          steps {
+            log(level: 'INFO', text: 'Launching Async ITs')
+            // TODO: AGENT_INTEGRATION_TEST = Opbeans when the apm-integration-testing gets merged
```
Comment (v1v) [inline .ci/Jenkinsfile]:
```suggestion
```

File: .ci/bump-version.sh [hunk 4]
```diff
@@ -0,0 +1,20 @@
+#!/usr/bin/env bash
+set -euxo pipefail
+
+if [ $# -lt 1 ]; then
+  echo "usage: ${0} folder"
```
Comment (mdelapenya) [inline .ci/bump-version.sh]: Is the message `Usage: ${0} folder` accurate?

File: .ci/bump-version.sh [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+  exit 1
+fi
+AGENT_VERSION="${1}"
+
+# Install Go using the same travis approach
```
Comment (kuisathaverat) [inline .ci/bump-version.sh]: probably we can put this script in the library https://github.com/elastic/beats/blob/b14f798d000f66582d37487bbda1fe9de6999e7d/.ci/scripts/install-go.sh in most of the places we use [gvm](https://github.com/andrewkroh/gvm)

Comment (v1v) [reply .ci/bump-version.sh]: The reason for using the travis approach was to align the same commands in travis and Jenkins for the apm-agent-go.

I did copy the logic from the apm-agent-go, so no strong opinion to change it and use something else as there is no requirements to use any specific approach in this repo.

File: .ci/bump-version.sh (hunk 4 above)
Comment (v1v) [inline .ci/bump-version.sh]:
```suggestion
  echo "usage: ${0} Go_Agent_Version"
```

File: .ci/bump-version.sh (hunk 5 above)
Comment (axw) [reply .ci/bump-version.sh]: The Go agent uses some features of gimme that are not available in gvm as far as I'm aware:
 - get the current `stable` release
 - get the most recent patch release of a specified minor (e.g. `1.12.x`)
 - build the current `master` from source

**Pull request merged**